### PR TITLE
Fix #10532

### LINKED ISSUE #10532: input's height is affected by parent's line-height
Test case:

``` html
<div style="line-height: 100px;">
    <input placeholder="foobar">
</div>
```

Firefox:
firefox

Servo:
servo

### PR DESCRIPTION
**Pull request by iamajoe**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10553)

### CODE REVIEW

**Comment by pcwalton** [comment]
Where does that line height value come from?

**Comment by Sendoushi** [comment]
Comparison with Firefox. Photoshop to check if everything is as supposed.

File: resources/servo.css [hunk 1]
```diff
@@ -1,5 +1,5 @@
-input                   { background: white; min-height: 1.0em; padding: 0em; padding-left: 0.25em; padding-right: 0.25em; border: solid lightgrey 1px; color: black; white-space: nowrap; text-align: left; }
-textarea                { background: white; min-height: 1.0em; padding: 0em; padding-left: 0.25em; padding-right: 0.25em; border: solid lightgrey 1px; color: black; white-space: pre; }
+input                   { background: white; min-height: 1.0em; padding: 0em; padding-left: 0.25em; padding-right: 0.25em; border: solid lightgrey 1px; color: black; font-family: sans-serif; font-size: 0.7em; white-space: nowrap; text-align: left; line-height: 1.8; }
+textarea                { background: white; min-height: 1.0em; padding: 0em; padding-left: 0.25em; padding-right: 0.25em; border: solid lightgrey 1px; color: black; font-family: sans-serif; font-size: 0.7em; white-space: pre; }
```
Comment (perlun) [inline resources/servo.css]: Shouldn't we try to split these properties onto multiple lines? It's awfully hard to meaningfully see the changes in the diff when the style is structured like that.

(I know it was the existing structure; I am just suggesting that we should clean it up. :smile:)

Comment (Sendoushi) [inline resources/servo.css]: I totally agree. I also find it weird to work with these.
Also... This file will definitely grow so, besides an import system we should maybe start commenting with sections? For example `/* Form */`.

**Comment by pcwalton** [comment]
Can you file a follow-up bug to get this added to the spec? https://html.spec.whatwg.org/multipage/rendering.html#form-controls

Once that's done, I'm fine with this patch.

**Comment by aethanyc** [comment]
Firefox use `line-height: normal` in [forms.css](https://dxr.mozilla.org/mozilla-central/rev/21bf1af375c1fa8565ae3bb2e89bd1a0809363d4/layout/style/res/forms.css#91l). 

Do we need a reftest to check these two pages are the same? These two look the same on Firefox. I'm not sure this is in the scope of this bug because even if we set the `line-height` of the `input` element, the line containing the `input` still have the height from the parent.
```html
<div style="line-height: 100px;">
  <input placeholder="foobar">
</div>
```
```html
<div>
  <input placeholder="foobar">
</div>
```

**Comment by Sendoushi** [comment]
@aethanyc Yes I noticed that `line-height: normal` in the firefox code before. The problem is that if I set with normal it won't get the right size. Setting line-height as I did, reseted to the right size but... maybe there is other way to set the default on servo so we could use that normal?

@pcwalton What would you want me to write? I'm not used to write these things :P New to contributing but want to keep on doing it!

**Comment by paulrouget** [comment]
(just a unrelated note: please don't name your PR "Fix #xxxx". It makes it a bit difficult to triage PRs and go through github notifications. And thank you for the PR ^^)

**Comment by Sendoushi** [comment]
@paulrouget Sure. How would you prefer me to name the PR? So that next time I've that in attention?

**Comment by aethanyc** [comment]
@Sendoushi I'm new to servo, too. Let's wait servo experts to answer your question.

**Comment by KiChjang** [comment]
@Sendoushi It'd be nice if you can have a descriptive title of what the PR is about. Coincidentally, your commit message for this PR is good enough as a PR title.

**Comment by jdm** [comment]
https://html.spec.whatwg.org/multipage/rendering.html#the-input-element-as-a-text-entry-widget is probably a more appropriate place in the specification to define this. The issue should be filed at https://github.com/whatwg/html/issues .

**Comment by perlun** [comment]
@jdm, @pcwalton - I think @Sendoushi would need some help here in what to write in the whatwg issue. Could either one of you (who know what to write :smile:) report the issue there and refer back to this PR?

**Comment by Sendoushi** [comment]
Yes please. I have no idea on the next steps...

**Comment by jdm** [comment]
I have filed https://github.com/whatwg/html/issues/1046 as requested.

**Comment by pcwalton** [comment]
Given that upstream spec states that this could reasonably be covered under the magic "binding" property, I think it's safe to say the spec allows us to do this today, so r=me.

**Comment by pcwalton** [comment]
Oh wait... r=me with a reftest added.

**Comment by Ms2ger** [comment]
(The spec problem is not so much "are we allowed to this" but "should we be *required* to do this")

**Comment by Sendoushi** [comment]
Here is the reftest: https://github.com/servo/servo/pull/10648 is it ok?

**Comment by notriddle** [comment]
There's no need to merge with master, or to have a separate pull request.

**Comment by Sendoushi** [comment]
This branch now has the tests may you please confirm that everything is as expected?

**Comment by notriddle** [comment]
You still have a bunch of merge commits in here (if you look at the commits tab, you can see all three of them). You can use git's rebase command to remove them.

**Comment by Sendoushi** [comment]
Oh. All in one commit. On it!

**Comment by Sendoushi** [comment]
Being hard to rebase this. Can anyone help?

**Comment by notriddle** [comment]
[Here's a nice article on how to do it, better than one I'd write probably](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html)

**Comment by perlun** [comment]
@Sendoushi - in this case you _should_ be able to do something like:

```
git fetch --all # Gives you the latest changes from all remotes
git rebase origin/master # This will replay your changes on top of current master, removing merge commits automatically
git push -f # Will force-push the branch to your fork
```

It shouldn't be harder than that. Give it a try and see.

**Comment by Sendoushi** [comment]
@notriddle checked that but it is hard to spot my commits now.
Tried that @perlun now it is even worse :s

**Comment by perlun** [comment]
What does `git remote -v` show? I think the problem may be that `origin` isn't the `servo` remote in this case, but your local fork, so you didn't rebase on top of what I expected.

Also, `git rebase -i origin/master` may be an idea where you can explicitly remove the lines for the commits that you don't want to include in the feature branch. Give it a try and see if you can make it work.

**Comment by Sendoushi** [comment]
@perlun yes, servo remote is "upstream". I tried -i but the problem with that is that i already have a lot of commits which aren't my own and it is pretty hard for me to go through each commit. I guess at this point I should just re-fork and re-do the whole thing.

**Comment by perlun** [comment]
Try 'git rebase upstream/master' and then force push. Should do the trick, no?

**Comment by Sendoushi** [comment]
That makes sense. Pulled upstream/master (everything up to date), rebase upstream/master it and conflict errors on `test-helpers.sub.js` and `MANIFEST.json`

**Comment by KiChjang** [comment]
Something is horribly wrong, not only do I see multiple commits, I see commits that are not from the PR author.

**Comment by perlun** [comment]
Weird. I'll see if I can sort it out.

**Comment by perlun** [comment]
@Sendoushi - I think I got the branch fixed. The steps I took:

1. Added your repo as a remote in my own working copy: `git remote add sendoushi https://github.com/Sendoushi/servo.git`
1. Fetched form all remotes, to ensure I had your changes and also the latest changes from servo upstream (the remote named `origin` in my case).
1. Checked out your branch and gave it a sensible name: `checkout -b sendoushi-master sendoushi/master`
1. Interactively rebased, to see the list of commits: `git rebase -i origin/master`. Gave me this output and no other commits:

```
pick 4cd043e [fix] Input line height and styling
pick db9244d [new] Input line height ref test
pick eff4835 [fix] Test for input in manifest
pick 27eea7b Improve properties.mako.rs file structure, take 2
```

The last one is a change I've made; I don't know how it made its way there. :smile: It's already merged into servo upstream, but I kept it anyway and let the rebase start.

I got an error when it attempted to apply that commit. Since I knew it's already merged and doesn't belong to your branch anyway, I skipped it in the rebase: `git rebase --skip`

The rebase was done. I pushed the result to my own fork: `git push perlun sendoushi-master`

And then I created the PR. Since I rebased your commits, they still have you as the author and you should get the credit you (rightfully) deserve.

Hope this helped! :smile:

**Comment by jdm** [comment]
Closing in favour of #10809. Thanks @perlun!

**Comment by Sendoushi** [comment]
@perlun it helped :) rebase is still a mysterious wizardry for me

**Comment by perlun** [comment]
@Sendoushi - it's one of the hardest parts of git perhaps, but don't be afraid: you can learn it as well. I happen to use it more or less every day in my work so I'm very used to the workflow.

(Think of rebase as "replay the commits from my current branch on top of some other commit than the commit I branched out from. Problems can occur if you rebase on top of the wrong commit etc, but once you learn it, it's one of the things with git that you tend to like the most.)

**Pull request closed without merging**